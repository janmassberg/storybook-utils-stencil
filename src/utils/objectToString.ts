import { toQuotedKeyName } from "./stringUtils";

/**
 * Converts any object to a string representation.
 * @param value {string|number|boolean|function|{}}
 */
export const objectToString = (value: any): string => {
    const valueType = typeof value;
    switch (valueType) {
        case "string":
            return `"${value}"`;
        case "number":
        case "boolean":
            return value.toString();
        case "function":
            return `function() {}`;
        case "object":
            if (Array.isArray(value)) {
                return `[${value.map(objectToString).join(", ")}]`;
            }
            if (value !== null) {
                const props = Object.entries(value).map(
                    ([key, value]) =>
                        `${toQuotedKeyName(key)}: ${objectToString(value)}`
                );
                return `{${props.join(", ")}}`;
            }
            return "";
        case "undefined":
        default:
            return "";
    }
};
