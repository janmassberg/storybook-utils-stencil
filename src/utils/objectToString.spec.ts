import { objectToString } from "./objectToString";

describe("objectToString", () => {
    it("returns strings with quotes", () => {
        expect(objectToString("test-string")).toEqual("\"test-string\"");
    });

    it("should convert numbers to strings", () => {
        expect(objectToString(1234)).toEqual("1234");
    });

    it("should convert boolean values to strings", () => {
        expect(objectToString(true)).toEqual("true");
        expect(objectToString(false)).toEqual("false");
    });

    it("should convert functions to strings", () => {
        expect(objectToString((event: Event) => {
            console.log(event);
        })).toMatchSnapshot();
    });

    it("should return empty string for undefined types", () => {
        expect(objectToString(undefined)).toEqual("");
        expect(objectToString(null)).toEqual("");
    });

    it("should convert objects to strings", () => {
        expect(objectToString({
            testString: "test string",
            testNumber: 1234,
            testTrue: true,
            testFalse: false,
        })).toMatchSnapshot();
    });
});
