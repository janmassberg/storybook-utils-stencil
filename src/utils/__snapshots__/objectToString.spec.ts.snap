// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`objectToString should convert functions to strings 1`] = `
"(event) => {
            console.log(event);
        }"
`;

exports[`objectToString should convert objects to strings 1`] = `"{testString: \\"test string\\", testNumber: 1234, testTrue: true, testFalse: false}"`;
